projects/online/online/main.py

```diff
@@ -245,26 +192,136 @@ def search(
 
         # if we found an event, process it!
         if event is not None:
-            logging.info("Found event")
-            process_event(
-                event,
-                gdb,
-                input_buffer,
-                spectral_density,
-                pe_whitener,
-                amplfi,
-                scaler,
-                pastro_model,
-                samples_per_event,
-                inference_params,
-                outdir,
-                nside,
-                device,
+            logging.info("Putting event in event queue")
+            event_queue.put(event)
+            logging.info("Running AMPLFI")
+            descaled_samples = run_amplfi(
+                event_time=event.gpstime,
+                input_buffer=input_buffer,
+                samples_per_event=samples_per_event,
+                spectral_density=spectral_density,
+                amplfi_whitener=amplfi_whitener,
+                amplfi=amplfi,
+                std_scaler=scaler,
+                device=device,
             )
+            shared_samples = descaled_samples.flatten()
+            amplfi_queue.put(event.gpstime)
             searcher.detecting = False
         # TODO write buffers to disk:
 
 
+def pastro_subprocess(
+    pastro_queue: Queue,
+    server: GdbServer,
+    background_path: Path,
+    foreground_path: Path,
+    rejected_path: Path,
+    astro_event_rate: float,
+    outdir: Path,
+):
+    gdb = gracedb_factory(server, outdir)
+
+    logging.info("Fitting p_astro model or loading from cache")
+    pastro_model = fit_or_load_pastro(
+        outdir / "pastro.pkl",
+        background_path,
+        foreground_path,
+        rejected_path,
+        astro_event_rate=astro_event_rate,
+    )
+    logging.info("Loaded p_astro model")
+
+    while True:
+        event = pastro_queue.get()
+        logging.info("Calculating p_astro")
+        pastro = pastro_model(event.detection_statistic)
+        graceid = pastro_queue.get()
+        logging.info(f"Submitting p_astro: {pastro}")
+        gdb.submit_pastro(float(pastro), graceid, event.gpstime)
+        logging.info("Submitted p_astro")
+
+
+def amplfi_subprocess(
+    amplfi_queue: Queue,
+    server: GdbServer,
+    outdir: Path,
+    inference_params: List[str],
+    shared_samples: Array,
+    nside: int = 32,
+):
+    gdb = gracedb_factory(server, outdir)
+
+    while True:
+        arg = amplfi_queue.get()
+        if isinstance(arg, float):
```

Comment:
Is this to also handle the local server? I think making the check on `server` will make it more obvious. Down the line we should also try to make the outputs of the `local` server more compatible with the real GDB servers too.

Reply:
No, this is to handle the fact that the `graceid` can get added to the `amplfi_queue` either before or after the `event_time`, depending on if AMPLFI finishes before or after event creation. I couldn't come up with a nice way to make sure that the arguments always entered in the same order.

Reply:
Ah interesting